### PR TITLE
feat: config based sdk

### PR DESCRIPTION
**Pull request by HarisChoudhary**

### CODE REVIEW

File: .threeport/sdk-config.yaml [hunk 1]
```diff
@@ -0,0 +1,211 @@
+APIObjects:
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:1]: APIObjectGroups

File: .threeport/sdk-config.yaml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  class:
+    - Name: Definition
+      Versions:
+        - v0
+      RouteExclude: true
+      DisableApiModel: true
+    - Name: Instance
+      Versions:
+        - v0
+      RouteExclude: true
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:79]: Merge this and below to route-exclude

Comment (HarisChoudhary) [reply .threeport/sdk-config.yaml:79]: ExcludeRoute

File: .threeport/sdk-config.yaml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    - Name: HelmWorkloadDefinition
+      Versions:
+        - v0
+      Reconcilable: true
+      TptctlConfigPath: true
+    - Name: HelmWorkloadInstance
+      Versions:
+        - v0
+      Reconcilable: true
+      TptctlConfigPath: true
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:116]: Merge with other tptctl to single struct:

type Tptctl struct {
	enabled  bool `yaml:",inline"`
	configPath bool
}

File: .threeport/sdk-config.yaml [hunk 4]
```diff
@@ -0,0 +1,211 @@
+APIObjects:
+  secret:
+    - Name: SecretDefinition
+      Versions:
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:4]: Change databasemarker DatabaseExclude to ExcludeFromDb

File: .threeport/sdk-config.yaml [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    - Name: ControlPlaneDefinition
+      Versions:
+        - v0
+      Reconcilable: true
+      Tptctl: true
+    - Name: ControlPlaneInstance
+      Versions:
+        - v0
+      Reconcilable: true
+      LoadAssociationsFromDatabase: true
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:62]: shorten database to db

File: .threeport/sdk-config.yaml [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        - v0
+      Reconcilable: true
+  actuator:
+    - Name: Profile
+      Versions:
+        - v0
+    - Name: Tier
+      Versions:
+        - v0
+  aws:
```
Comment (HarisChoudhary) [inline .threeport/sdk-config.yaml:20]: # aws:
- Name: aws
  Objects:

File: cmd/sdk/cmd/gen/generate.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/spf13/cobra"
+	"github.com/threeport/threeport/cmd/sdk/cmd"
+	"github.com/threeport/threeport/internal/sdk"
+)
+
+// genCmd represents the parent command for all codegen related operations
+var genCmd = &cobra.Command{
+	Use:   "gen",
+	Short: "Generate code for Threeport or its extensions.",
+	Long:  `Generate code for Threeport or its extensions.`,
```
Comment (HarisChoudhary) [inline cmd/sdk/cmd/gen/generate.go:18]: Make this more descriptive. Synthesize summary for other three commands

File: cmd/sdk/cmd/create/api_objects.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	apiobjectmanager "github.com/threeport/threeport/internal/sdk/api-object-manager"
+	cli "github.com/threeport/threeport/pkg/cli/v0"
+)
+
+var createAPIObjectConfig string
+
+// createAPICmd represents the cmd to create an api-object with the sdk in threeport
+var createAPICmd = &cobra.Command{
+	Use:   "api-objects",
+	Short: "Create threeport api objects.",
```
Comment (HarisChoudhary) [inline cmd/sdk/cmd/create/api_objects.go:22]: Create initial scaffloding for api objects

File: cmd/sdk/cmd/create/create.go [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"github.com/spf13/cobra"
+	"github.com/threeport/threeport/cmd/sdk/cmd"
+)
+
+// createCmd represents the createCmd command
+var createCmd = &cobra.Command{
+	Use:   "create",
+	Short: "Create aspects of the Threeport API or its extensions.",
```
Comment (HarisChoudhary) [inline cmd/sdk/cmd/create/create.go:14]: Create initial source code scaffolding

**Review by randalljohnson — APPROVED** [review]
🚀

**Comment by lander2k2** [comment]
Suggested change on lines 40 and 47 of `internal/sdk/models/command.go':
```
...HeaderComment("originally generated by 'threeport-sdk codegen api-model' but will not be regenerated - intended for modification")
```
Should be:
```
...HeaderComment("originally generated by 'threeport-sdk codegen' for tptctl command scaffolding but will not be regenerated - intended for modification")
```

File: internal/sdk/sdk_config.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+	// Tptctl contains sdk configurations related to tptctl
+	Tptctl *Tptctl `yaml:"Tptctl"`
+}
+
+type APIObjectConfig struct {
+	SDKConfig `yaml:",inline"`
+}
+
+type Tptctl struct {
```
Comment (lander2k2) [inline internal/sdk/sdk_config.go:70]:
```suggestion
// Tptctl contains attributes used by the SDK to generate tptctl
// command source code.
type Tptctl struct {
```

**Review by lander2k2 — CHANGES_REQUESTED** [review]

**Comment by lander2k2** [comment]
Suggest changing the `APIObjectManager` struct name to `CodegenManager` in `internal/sdk/api-object-manager/manager.go` to better align with its intended future purpose.

Would also make sense to change the directory from `api-object-manager` to `codegen-manager` and the package name from `apiobjectmanager` to `codegenmanager` as well.

**Comment by lander2k2** [comment]
Suggest changing the `SDKConfig` object and `APIObjectConfig` struct names.  And rearrange the ordering of the structs to represent the correct hierachy:

```
// was: type APIObjectConfig struct {
type SdkConfig struct {
	ApiObjectConfig `yaml:",inline"`
}

// APIObjectConfig contains the config for the threeport sdk to use
// It is a map of controller domains and the api objects under them
//was: type SDKConfig struct {
type ApiObjectConfig struct {
	ApiObjectGroups []*APIObjectGroup `yaml:"APIObjectGroups"`
}
```

**Comment by lander2k2** [comment]
For naming conventions in `internal/sdk/sdk_config.go` and `internal/sdk/api-object-manager/manager.go`, can we remove the all-caps for acronyms as per our [style guide](https://github.com/threeport/threeport/blob/main/docs/style-guide.md#naming)?

So `APIObject` becomes `ApiObject` etc.

File: internal/sdk/sdk_config.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	SDKConfigType = "yaml"
+)
+
+// SDKConfig contains the config for the threeport sdk to use
+// It is a map of controller domains and the api objects under them
+type SDKConfig struct {
+	APIObjectGroups []*APIObjectGroup `yaml:"APIObjectGroups"`
+}
+
+type APIObjectGroup struct {
```
[the remark this answers is not in the document]
Comment (lander2k2) [reply internal/sdk/sdk_config.go:23]: Note: there is another comment that addresses this section of code: https://github.com/threeport/threeport/pull/258#issuecomment-2023818410

**Review by lander2k2 — APPROVED** [review]

**Pull request merged**